clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -498,6 +498,104 @@ public void validateOffsetsIfNeeded() {
         validateOffsetsAsync(partitionsToValidate);
     }
 
+    /**
+     * For each partition which needs validation, make an asynchronous request to get the end-offsets for the partition
+     * with the epoch less than or equal to the epoch the partition last saw.
+     *
+     * Requests are grouped by Node for efficiency.
+     */
+    private void validateOffsetsAsync(Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate) {
+        final Map<Node, Map<TopicPartition, SubscriptionState.FetchPosition>> regrouped =
+            regroupFetchPositionsByLeader(partitionsToValidate);
+
+        regrouped.forEach((node, fetchPositions) -> {
+            if (node.isEmpty()) {
+                metadata.requestUpdate();
+                return;
+            }
+
+            NodeApiVersions nodeApiVersions = apiVersions.get(node.idString());
+            if (nodeApiVersions == null) {
+                client.tryConnect(node);
+                return;
+            }
+
+            if (!hasUsableOffsetForLeaderEpochVersion(nodeApiVersions)) {
+                log.debug("Skipping validation of fetch offsets for partitions {} since the broker does not " +
+                              "support the required protocol version (introduced in Kafka 2.3)",
+                    fetchPositions.keySet());
+                completeAllValidations(fetchPositions);
+                return;
+            }
+
+            // We need to get the client epoch state before sending out the leader epoch request, and use it to
+            // decide whether we need to validate offsets.
+            if (!metadata.hasReliableLeaderEpochs()) {
```

Comment:
Change one: immediately complete the validation when the leader epoch is not reliable.

Reply:
Could you not move the function while changing it so that the diff is easier to check -- e.g. here I'd have to blindly trust you that there are only changes of this func :)

Reply:
Np, will do